File: engine/language-server/src/main/scala/org/enso/languageserver/boot/resource/DirectoriesInitialization.scala
package org.enso.languageserver.boot.resource

import com.typesafe.scalalogging.LazyLogging
import org.enso.languageserver.data.ProjectDirectoriesConfig

import scala.concurrent.{ExecutionContext, Future}

/** Directories initialization.
  *
  * @param directoriesConfig the directories config
  */
final class DirectoriesInitialization(
  directoriesConfig: ProjectDirectoriesConfig
)(implicit
  ec: ExecutionContext
) extends InitializationComponent
    with LazyLogging {

  @volatile
  private var _isInitialized: Boolean = false

  /** @inheritdoc */
  override def isInitialized: Boolean = _isInitialized

  /** @inheritdoc */
  override def init(): Future[InitializationComponent.Initialized.type] = {
    if (isInitialized) Future.successful(InitializationComponent.Initialized)
    else
      Future {
        logger.info("Initializing directories...")
        directoriesConfig.createDirectories()
        logger.info("Initialized directories.")
        _isInitialized = true
        InitializationComponent.Initialized
      }
  }
}

File: engine/language-server/src/main/java/org/enso/languageserver/boot/resource/DirectoriesInitialization.java
package org.enso.languageserver.boot.resource;

import java.util.concurrent.Executor;
import org.enso.languageserver.data.ProjectDirectoriesConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Directories initialization. */
public class DirectoriesInitialization extends LockedInitialization {
  private final ProjectDirectoriesConfig projectDirectoriesConfig;
  private final Logger logger = LoggerFactory.getLogger(this.getClass());

  /**
   * Creates the directories initialization component.
   *
   * @param executor the executor that runs the initialization
   * @param projectDirectoriesConfig the directories config
   */
  public DirectoriesInitialization(
      Executor executor, ProjectDirectoriesConfig projectDirectoriesConfig) {
    super(executor);
    this.projectDirectoriesConfig = projectDirectoriesConfig;
  }

  @Override
  public void initComponent() {
    logger.info("Initializing directories...");
    projectDirectoriesConfig.createDirectories();
    logger.info("Initialized directories.");
  }
}
